internal/models/openai/openai_embedding.go

```diff
@@ -0,0 +1,238 @@
+	User string `json:"user,omitempty"`
+
+	// The format to return the embeddings in. Can be either float or base64.
+	EncodingFormat string `json:"encoding_format,omitempty"`
+
+	// The number of dimensions the resulting output embeddings should have. Only supported in text-embedding-3 and later models.
+	Dimensions int `json:"dimensions,omitempty"`
+}
+
+type Usage struct {
+	// The number of tokens used by the prompt.
+	PromptTokens int `json:"prompt_tokens"`
+
+	// The total number of tokens used by the request.
+	TotalTokens int `json:"total_tokens"`
+}
+
+type EmbeddingData struct {
+	// The object type, which is always "embedding".
+	Object string `json:"object"`
+
+	// The embedding vector, which is a list of floats.
+	Embedding []float32 `json:"embedding"`
+
+	// The index of the embedding in the list of embeddings.
+	Index int `json:"index"`
+}
+
+type EmbeddingResponse struct {
+	// The object type, which is always "list".
+	Object string `json:"object"`
+
+	// The list of embeddings generated by the model.
+	Data []EmbeddingData `json:"data"`
+
+	// The name of the model used to generate the embedding.
+	Model string `json:"model"`
+
+	// The usage information for the request.
+	Usage Usage `json:"usage"`
+}
+
+type ByIndex struct {
+	resp *EmbeddingResponse
+}
+
+func (eb *ByIndex) Len() int { return len(eb.resp.Data) }
+func (eb *ByIndex) Swap(i, j int) {
+	eb.resp.Data[i], eb.resp.Data[j] = eb.resp.Data[j], eb.resp.Data[i]
+}
+func (eb *ByIndex) Less(i, j int) bool { return eb.resp.Data[i].Index < eb.resp.Data[j].Index }
+
+type ErrorInfo struct {
+	Code    string `json:"code"`
+	Message string `json:"message"`
+	Param   string `json:"param,omitempty"`
+	Type    string `json:"type"`
+}
+
+type EmbedddingError struct {
+	Error ErrorInfo `json:"error"`
+}
+
+type OpenAIEmbeddingInterface interface {
+	Check() error
+	Embedding(modelName string, texts []string, dim int, user string, timeoutSec time.Duration) (*EmbeddingResponse, error)
+}
+
+type openAIBase struct {
+	apiKey string
+	url    string
+}
+
+func (c *openAIBase) Check() error {
+	if c.apiKey == "" {
+		return fmt.Errorf("api key is empty")
+	}
+
+	if c.url == "" {
+		return fmt.Errorf("url is empty")
+	}
+	return nil
+}
+
+func (c *openAIBase) genReq(modelName string, texts []string, dim int, user string) *EmbeddingRequest {
+	var r EmbeddingRequest
+	r.Model = modelName
+	r.Input = texts
+	r.EncodingFormat = "float"
+	if user != "" {
+		r.User = user
+	}
+	if dim != 0 {
+		r.Dimensions = dim
+	}
+	return &r
+}
+
+type OpenAIEmbeddingClient struct {
+	openAIBase
+}
+
+func NewOpenAIEmbeddingClient(apiKey string, url string) *OpenAIEmbeddingClient {
+	return &OpenAIEmbeddingClient{
+		openAIBase{
+			apiKey: apiKey,
+			url:    url,
+		},
+	}
+}
+
+func (c *OpenAIEmbeddingClient) Embedding(modelName string, texts []string, dim int, user string, timeoutSec time.Duration) (*EmbeddingResponse, error) {
```

Comment:
this function is largely duplicate with `func (c *AzureOpenAIEmbeddingClient) Embedding`, can we try share the same logic?